File: app/src/main/java/ch/threema/app/locationpicker/PoiRepository.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2019-2025 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.locationpicker;

import android.annotation.SuppressLint;
import android.os.AsyncTask;
import android.widget.Toast;

import org.maplibre.android.geometry.LatLng;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.slf4j.Logger;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

import javax.net.ssl.HttpsURLConnection;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import ch.threema.app.ThreemaApplication;
import ch.threema.app.managers.ServiceManager;
import ch.threema.app.utils.ConfigUtils;
import ch.threema.app.utils.RuntimeUtil;
import ch.threema.app.utils.TestUtil;
import ch.threema.base.ThreemaException;
import ch.threema.base.utils.LoggingUtil;
import ch.threema.domain.protocol.ProtocolStrings;

class PoiRepository {
    private static final Logger logger = LoggingUtil.getThreemaLogger("PoiRepository");
    public static final int QUERY_MIN_LENGTH = 3;

    private List<Poi> places = new ArrayList<>();
    private final MutableLiveData<List<Poi>> mutableLiveData = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>();

    @SuppressLint("StaticFieldLeak")
    MutableLiveData<List<Poi>> getMutableLiveData(PoiQuery poiQuery) {
        logger.debug("getMutableLiveData");

        new AsyncTask<Void, Void, Void>() {
            @Override
            protected void onPreExecute() {
                isLoading.setValue(true);
            }

            @Override
            protected Void doInBackground(Void... voids) {
                places.clear();
                final var query = poiQuery.getQuery();
                final var center = poiQuery.getCenter();

                if (query == null || query.length() < QUERY_MIN_LENGTH || center == null) {
                    return null;
                }
                if (center.getLatitude() == 0.0d && center.getLongitude() == 0.0d) {
                    return null;
                }

                try {
                    final ServiceManager serviceManager = ThreemaApplication.getServiceManager();
                    if (serviceManager == null) {
                        logger.error("Could not obtain service manager");
                        return null;
                    }
                    @NonNull
                    final var addressProvider = serviceManager.getServerAddressProviderService().getServerAddressProvider();

                    final String placesUrl = addressProvider.getMapPOINamesUrl().get(center.getLatitude(), center.getLongitude(), query);
                    URL serverUrl = new URL(placesUrl);
                    logger.debug("Places URL: {}", placesUrl);
                    HttpsURLConnection urlConnection = null;
                    try {
                        urlConnection = (HttpsURLConnection) serverUrl.openConnection();
                        urlConnection.setSSLSocketFactory(ConfigUtils.getSSLSocketFactory(serverUrl.getHost()));
                        urlConnection.setConnectTimeout(15000);
                        urlConnection.setReadTimeout(30000);
                        urlConnection.setRequestProperty("User-Agent", ProtocolStrings.USER_AGENT);
                        urlConnection.setRequestMethod("GET");
                        urlConnection.setDoOutput(false);

                        int responseCode = urlConnection.getResponseCode();

                        if (responseCode == HttpsURLConnection.HTTP_OK) {
                            StringBuilder sb = new StringBuilder();
                            try (BufferedReader br = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()))) {
                                String line;
                                while ((line = br.readLine()) != null) {
                                    sb.append(line).append("\n");
                                }
                            }
                            try {
                                parseJson(sb.toString());
                            } catch (JSONException e) {
                                logger.error("Exception", e);
                            }
                        } else {
                            logger.info("Unable to fetch POI names: {}", urlConnection.getResponseMessage());
                        }
                    } finally {
                        if (urlConnection != null) {
                            urlConnection.disconnect();
                        }
                    }
                } catch (IOException | ThreemaException e) {
                    logger.error("Exception", e);
                }

                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                isLoading.setValue(false);
                mutableLiveData.setValue(places);
            }
        }.execute();

        return mutableLiveData;
    }

    MutableLiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    private void parseJson(@NonNull String json) throws JSONException {
        JSONArray jsonArray = new JSONArray(json);
        if (jsonArray.length() > 0) {
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject result = jsonArray.getJSONObject(i);

                if (result != null) {
                    Poi place = new Poi();
                    double lat = result.getDouble("lat");
                    double lng = result.getDouble("lon");
                    place.setLatLng(new LatLng(lat, lng));
                    place.setName(result.getString("name"));

                    String placeS = result.optString("place");
                    String highway = result.optString("highway");

                    if (result.has("dist")) {
                        place.setDistance(result.getInt("dist"));
                    } else {
                        place.setDistance(-1);
                    }

                    if (!TestUtil.isEmptyOrNull(highway)) {
                        place.setDescription("street");
                    } else {
                        place.setDescription(placeS);
                    }

                    places.add(place);
                }
            }
        }
    }
}

File: app/src/main/java/ch/threema/app/location/PoiRepository.kt
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2025 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.location

import ch.threema.base.ThreemaException
import ch.threema.common.buildRequest
import ch.threema.common.executeAsync
import ch.threema.common.getStringOrNull
import ch.threema.common.getSuccessBodyOrThrow
import ch.threema.common.models.Coordinates
import ch.threema.common.toJSONObjectList
import ch.threema.domain.protocol.ServerAddressProvider
import java.io.IOException
import okhttp3.OkHttpClient
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class PoiRepository(
    private val okHttpClient: OkHttpClient,
    private val serverAddressProvider: ServerAddressProvider,
) {
    @Throws(ThreemaException::class, IOException::class)
    suspend fun getNearbyPois(
        center: Coordinates,
        limit: Int,
    ): List<NearbyPoi> {
        val parameterizedUrl = serverAddressProvider.getMapPoiAroundUrl()
            ?: throw ThreemaException("No URL available to fetch nearby POIs")
        val request = buildRequest {
            url(parameterizedUrl.get(center, NEARBY_RADIUS))
        }

        val response = okHttpClient.executeAsync(request)
        try {
            val results = JSONArray(response.getSuccessBodyOrThrow().string())
            return results.toJSONObjectList()
                .mapNotNull { result ->
                    parseNearbyPoiResult(result)
                }
                .take(limit)
        } catch (e: JSONException) {
            throw ThreemaException("Failed to parse nearby POI response", e)
        }
    }

    private fun parseNearbyPoiResult(result: JSONObject): NearbyPoi? {
        return NearbyPoi(
            id = result.getLong("id"),
            name = result.getString("name"),
            coordinates = Coordinates(
                latitude = result.getDouble("lat"),
                longitude = result.getDouble("lon"),
            ),
            type = parseNearbyPoiType(result) ?: return null,
            isNatural = result.has("natural"),
        )
    }

    private fun parseNearbyPoiType(result: JSONObject): String? {
        val type = result.getStringOrNull("amenity")
            ?: result.getStringOrNull("shop")
            ?: result.getStringOrNull("tourism")
            ?: result.getStringOrNull("sport")
            ?: result.getStringOrNull("leisure")
            ?: result.getStringOrNull("natural")
            ?: result.getStringOrNull("public_transport")
            ?: (if (result.has("aeroway")) "airport" else null)
            ?: (if (result.has("aerialway")) "cablecar" else null)
            ?: result.getStringOrNull("highway")

        return when (type) {
            "fast_food",
            "food_court",
            -> "food"
            "hairdresser",
            -> "hair_care"
            "fitness_centre",
            -> "gym"
            "do_it_yourself",
            "doityourself",
            -> "hardware_store"
            "kindergarten",
            -> "child_care"
            "nursing_home",
            "clinic",
            "social_facility",
            -> "hospital"
            "parking_entrace",
            -> "parking"
            "playground",
            -> "park"
            "car_sharing",
            -> "car_rental"
            "car",
            -> "car_dealer"
            "hotel",
            -> "lodging"
            "doctors",
            -> "doctor"
            "recycling",
            -> "establishment"
            "mall",
            -> "shopping_mall"
            "swimming_pool",
            "water_park",
            -> "swimming"
            "arts_centre",
            -> "gallery"
            "dry_cleaning",
            -> "laundry"
            "second_hand",
            -> "furniture"
            "boutique",
            "clothes",
            -> "clothing_store"
            "beauty",
            "depilation",
            -> "beauty_salon"
            "chemist",
            -> "health"
            "townhall",
            -> "local_government_office"
            "bicycle",
            "bicycle_rental",
            -> "bicycle_store"
            "cave_entrance",
            -> "park"
            "peak",
            "volcano",
            -> "mountain"
            "pub",
            -> "bar"
            "stop_position",
            -> "bus_station"
            "stationery",
            "kiosk",
            -> "store"
            "shoes",
            -> "shoe_store"
            "aerialway",
            -> "cablecar"
            "rest_area",
            "services",
            -> "highway"
            "cinema",
            -> "movie_theater"
            else -> type
        }
    }

    @Throws(ThreemaException::class, IOException::class)
    suspend fun getPoiNames(
        center: Coordinates,
        query: String,
    ): List<NamedPoi> {
        val parameterizedUrl = serverAddressProvider.getMapPoiNamesUrl()
            ?: throw ThreemaException("No URL available to fetch POI names")
        val request = buildRequest {
            url(parameterizedUrl.get(center, query))
        }

        val response = okHttpClient.executeAsync(request)
        try {
            val results = JSONArray(response.getSuccessBodyOrThrow().string())
            return results.toJSONObjectList()
                .mapNotNull { result ->
                    parseNamedPoiResult(result)
                }
        } catch (e: JSONException) {
            throw ThreemaException("Failed to parse POI names response", e)
        }
    }

    private fun parseNamedPoiResult(result: JSONObject): NamedPoi? {
        return NamedPoi(
            name = result.getString("name"),
            coordinates = Coordinates(
                latitude = result.getDouble("lat"),
                longitude = result.getDouble("lon"),
            ),
            distance = if (result.has("dist")) {
                result.getInt("dist")
            } else {
                return null
            },
            description = if (result.has("highway")) {
                "street"
            } else {
                result.getStringOrNull("place")
            },
        )
    }

    companion object {
        /**
         * Search radius in meters
         */
        private const val NEARBY_RADIUS = 750
    }
}
